lib/executor.js

```diff
@@ -330,24 +334,42 @@ function setupMiddleware(app, instructions) {
     }
     assert(typeof factory === 'function',
       'Middleware factory must be a function');
-    data.config = getUpdatedConfigObject(app, data.config);
+    var opts = {
+      useEnvVars: true,
+    };
+    data.config = getUpdatedConfigObject(app, data.config, opts);
     app.middlewareFromConfig(factory, data.config);
   });
 }
 
-function getUpdatedConfigObject(app, config) {
+function getUpdatedConfigObject(app, config, opts) {
   var DYNAMIC_CONFIG_PARAM = /\$\{(\w+)\}$/;
+  var useEnvVars = opts.useEnvVars || undefined;
 
   function getConfigVariable(param) {
     var configVariable = param;
     var match = configVariable.match(DYNAMIC_CONFIG_PARAM);
     if (match) {
-      var appValue = app.get(match[1]);
-      if (appValue !== undefined) {
+      var varName = match[1];
+      if (useEnvVars && process.env[varName] !== undefined) {
+        debug('Dynamic Configuration: Resolved via process.env: %s as %s',
+          process.env[varName], param);
+        configVariable = process.env[varName];
+      } else if (app.get(varName) !== undefined) {
+        debug('Dynamic Configuration: Resolved via app.get(): %s as %s',
+          app.get(varName), param);
+        var appValue = app.get(varName);
         configVariable = appValue;
       } else {
-        console.warn('%s does not resolve to a valid value. ' +
-        '"%s" must be resolvable by app.get().', param, match[1]);
+        // previously it returns the original string such as "${restApiRoot}"
+        // it will now return `undefined`, for the use case of
+        // dynamic datasources url:`undefined` to fallback to other parameters
+        configVariable = undefined;
+        console.warn('%s does not resolve to a valid value, returned as %s. ' +
+        '"%s" must be resolvable in Environment variable or by app.get().',
+          param, configVariable, param);
```

Comment:
In this case 
`param` would be `${restApiRoot}`
and `varName` would be `restApiRoot`, I feel either one would make sense. Will update it